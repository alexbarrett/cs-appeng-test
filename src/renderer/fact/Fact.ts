export interface Fact {
  source: string;
  id: string;
  text: string;
  updatedAt: string;
}

export const isFact = (value: any): value is Fact =>
  typeof value === 'object' &&
  ['source', 'id', 'text', 'updatedAt'].every(
    (property) => typeof value[property] === 'string',
  );

export type FactDataSource = () => Promise<Fact>;
