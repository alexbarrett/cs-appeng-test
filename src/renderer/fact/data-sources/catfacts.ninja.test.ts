import fetchMock from 'jest-fetch-mock';

import { FactDataSource } from '../Fact';
import { createDataSource } from './catfacts.ninja';

describe('data source', () => {
  let loadFact: FactDataSource;

  beforeAll(() => {
    loadFact = createDataSource();

    fetchMock.mockResponse(
      JSON.stringify({
        fact: 'Researchers believe the word “tabby” comes from Attabiyah, a neighborhood in Baghdad, Iraq. Tabbies got their name because their striped coats resembled the famous wavy patterns in the silk produced in this city.',
        length: 212,
      }),
    );
  });

  it('is able to parse a valid response from the API', async () => {
    const fact = await loadFact();
    expect(fact).toMatchObject({
      text: expect.stringContaining('tabby'),
      updatedAt: expect.stringMatching(/^[\d-]+T[\d:]+/), // ISO timestamp-ish.
      source: expect.any(String),
    });
  });

  it('generates an ID from the fact text', async () => {
    const fact = await loadFact();
    expect(fact.id).toMatchSnapshot(); // Unable to use inline snapshots with Prettier v3 yet.
  });
});
