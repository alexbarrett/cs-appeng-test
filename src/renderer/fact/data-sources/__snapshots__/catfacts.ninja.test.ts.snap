// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`data source generates an ID from the fact text 1`] = `"PoJZ1znsnFhKKYpxaNIYP7Gj9Yo="`;
